plutus-core-interpreter/src/Language/PlutusCore/Interpreter/LMachine.hs

```diff
@@ -0,0 +1,278 @@
+-- | The L machine
+-- A lazy machine based on the L machine of Friedman et al. [Improving the Lazy Krivine Machine]
+-- For more details see the document in plutus/docs/fomega/lazy-machine
+-- The code here's closely based on the CEK machine implementation.
+
+-- The main difference from the CEK machine is that we have a heap
+-- containing closures containing evaluated/unevaluated function
+-- arguments, and environments mapping variable IDs to heap locations.
+-- To evaluate [M N], we save a closure C containing N and the current
+-- environment in a stack frame, then evaluate M to get a value
+-- (lambda/builtin) like \x.e and an environment env.  We retrieve the
+-- argument closure C from the stack, generate a new heap location l,
+-- bind C (unevaluated) to l in the heap, then evaluate e in
+-- env[x->l].  When we come across a use of x inside e, we look up l
+-- in the heap to get N (inside the closure C): if it's Evaluated then
+-- we continue immediately, using N instead of x. Otherwise we place
+-- an update marker on the stack and evaluate N to get a new closure
+-- C' containing a value V (and a possibly new environment): by
+-- this time we should be back at the update marker containing l, and
+-- we store Evaluated C back in the heap at l, then continue
+-- evaluating e.
+
+
+module Language.PlutusCore.Interpreter.LMachine
+    ( EvaluationResultF (EvaluationSuccess, EvaluationFailure)
+    , EvaluationResult
+    , evaluateL
+    , runL
+    ) where
+
+import           Language.PlutusCore
+import           Language.PlutusCore.Constant
+import           Language.PlutusCore.View
+import           PlutusPrelude
+
+import           Control.Monad.Identity
+import           Data.IntMap                  (IntMap)
+import qualified Data.IntMap                  as IntMap
+
+type Plain f = f TyName Name ()
+
+-- Let's just throw exceptions for the time being: error handling is changeable at the moment
+data LMachineException =
+    LMachineException String (Plain Term)
+  | LMachineStringException String
```

Comment:
Having you tried reusing the machinery from `Language.PlutusCore.Evaluation.MachineException`? It provides common errors that can occur during evaluation, but also allows to extend that set of errors. Both the CK and the CEK machines use it and extend it with their own exceptions. You'd have to add exceptions related to heap and the dynamic built-in names related exception from the CEK machine (we should abstract this stuff, so it can be reused across machines).

Reply:
Yes, I'm in the process of trying to fix that at the moment.  My exceptions are escaping to the outside world and causing CI failures.  I did use these earlier but took them out or some reason, I think when I was trying to make it monadic.

Reply:
The CI fails not because you throw something the tests do not know about, but because you throw where you shouldn't throw at all. See [this comment](https://github.com/input-output-hk/plutus/pull/331#issuecomment-441248679).

Reply:
FIxed.